### DISCUSSION TITLE
Error while using ma27 solver

### DISCUSSION THREAD

**Discussion by wissamkafa** [casadi/casadi#3291] [Q&A]
Hi, 
I have a python program that uses CasADi and IPOPT.
The code runs as required sith mumps solvers, but when i tried to use ma27, i have some problems.
I get the HSL packages from their offecial website.
extracted and compiled them follwoing the steps in CasADi
https://github.com/casadi/casadi/wiki/Obtaining-HSL
"
extraction folder: /home/user/ThirdParty-HSL/coinhsl

1)  ./configure --prefix=/usr/local LIBS="-llapack" --with-blas="-L/usr/lib -lblas" CXXFLAGS="-g -O2 -fopenmp" FCFLAGS="-g -O2 -fopenmp" CFLAGS="-g -O2 -fopenmp"
2) make
3) make install
"

then I created the symbolic link to obtain libhsl.so as required
 ln -s /usr/local/lib/libcoinhsl.so /usr/local/lib/libhsl.so

 then to make IOPOT see the solvers i added the environment variable using
 export LD_LIBRARY_PATH="$LD_LIBRARY_PATH:/usr/local/lib"
when using: echo $LD_LIBRARY_PATH i get:
:/usr/local/lib:/home/user/ThirdParty-HSL/coinhsl/lib

 now when using ('nlpsol_opts': {'ipopt.linear_solver': 'MA27'}) within the code to use ma27 i got them message:
Exception of type: DYNAMIC_LIBRARY_FAILURE in file "/work/build/external_projects/src/ipopt-external/src/Common/IpLibraryLoader.cpp" at line 72:
 Exception message: libhsl.so: cannot open shared object file: No such file or directory
 EXIT: Library loading failure.
 
 It seems that the paths could be found but it could not load the library. I don't know whya and I almost tried everything, but did not work.
 
 I am using ubuntu 22.04
 python 3.10 - Anaconda
 pycharm environment
 CasADi 3.6.3
 and I think all required packages and compilers are installed as required.
 
 I have been working to overcome this problem but with no success, and it is now driving me craizy
 
 
 could you please advise me what to do? and if some one has similar problem how he solved it?

**Comment by jgillis** [comment]
You may run your code with `LD_DEBUG=libs` to check where the dynamic loader fails..

**Comment by wissamkafa** [comment]
Thanks for your reply.
Actually, I did that and got the output:


     29903:	find library=libhsl.so [0]; searching
     29903:	 search path=/usr/local/lib/python3.10/dist-packages/casadi:/usr/local/lib/python3.10/dist-packages/casadi/./glibc-hwcaps/x86-64-v3:/usr/local/lib/python3.10/dist-packages/casadi/./glibc-hwcaps/x86-64-v2:/usr/local/lib/python3.10/dist-packages/casadi/./tls/haswell/x86_64:/usr/local/lib/python3.10/dist-packages/casadi/./tls/haswell:/usr/local/lib/python3.10/dist-packages/casadi/./tls/x86_64:/usr/local/lib/python3.10/dist-packages/casadi/./tls:/usr/local/lib/python3.10/dist-packages/casadi/./haswell/x86_64:/usr/local/lib/python3.10/dist-packages/casadi/./haswell:/usr/local/lib/python3.10/dist-packages/casadi/./x86_64:/usr/local/lib/python3.10/dist-packages/casadi/.		(RPATH from file /usr/local/lib/python3.10/dist-packages/casadi/_casadi.so)
     29903:	  trying file=/usr/local/lib/python3.10/dist-packages/casadi/libhsl.so
     29903:	  trying file=/usr/local/lib/python3.10/dist-packages/casadi/./glibc-hwcaps/x86-64-v3/libhsl.so
     29903:	  trying file=/usr/local/lib/python3.10/dist-packages/casadi/./glibc-hwcaps/x86-64-v2/libhsl.so
     29903:	  trying file=/usr/local/lib/python3.10/dist-packages/casadi/./tls/haswell/x86_64/libhsl.so
     29903:	  trying file=/usr/local/lib/python3.10/dist-packages/casadi/./tls/haswell/libhsl.so
     29903:	  trying file=/usr/local/lib/python3.10/dist-packages/casadi/./tls/x86_64/libhsl.so
     29903:	  trying file=/usr/local/lib/python3.10/dist-packages/casadi/./tls/libhsl.so
     29903:	  trying file=/usr/local/lib/python3.10/dist-packages/casadi/./haswell/x86_64/libhsl.so
     29903:	  trying file=/usr/local/lib/python3.10/dist-packages/casadi/./haswell/libhsl.so
     29903:	  trying file=/usr/local/lib/python3.10/dist-packages/casadi/./x86_64/libhsl.so
     29903:	  trying file=/usr/local/lib/python3.10/dist-packages/casadi/./libhsl.so
     29903:	 search path=/usr/local/lib/python3.10/dist-packages/casadi:/usr/local/lib/python3.10/dist-packages/casadi/.		(RPATH from file /usr/local/lib/python3.10/dist-packages/casadi/_casadi.so)
     29903:	  trying file=/usr/local/lib/python3.10/dist-packages/casadi/libhsl.so
     29903:	  trying file=/usr/local/lib/python3.10/dist-packages/casadi/./libhsl.so
     29903:	 search path=glibc-hwcaps/x86-64-v3:glibc-hwcaps/x86-64-v2:tls/haswell/x86_64:tls/haswell:tls/x86_64:tls:haswell/x86_64:haswell:x86_64::/home/user/ThirdParty-HSL/coinhsl/lib		(LD_LIBRARY_PATH)
     29903:	  trying file=glibc-hwcaps/x86-64-v3/libhsl.so
     29903:	  trying file=glibc-hwcaps/x86-64-v2/libhsl.so
     29903:	  trying file=tls/haswell/x86_64/libhsl.so
     29903:	  trying file=tls/haswell/libhsl.so
     29903:	  trying file=tls/x86_64/libhsl.so
     29903:	  trying file=tls/libhsl.so
     29903:	  trying file=haswell/x86_64/libhsl.so
     29903:	  trying file=haswell/libhsl.so
     29903:	  trying file=x86_64/libhsl.so
     29903:	  trying file=libhsl.so
     29903:	  trying file=/home/user/ThirdParty-HSL/coinhsl/lib/libhsl.so
     29903:	
     29903:	find library=libgfortran.so.5 [0]; searching
     29903:	 search path=glibc-hwcaps/x86-64-v3:glibc-hwcaps/x86-64-v2:tls/haswell/x86_64:tls/haswell:tls/x86_64:tls:haswell/x86_64:haswell:x86_64::/home/user/ThirdParty-HSL/coinhsl/lib		(LD_LIBRARY_PATH)
     29903:	  trying file=glibc-hwcaps/x86-64-v3/libgfortran.so.5
     29903:	  trying file=glibc-hwcaps/x86-64-v2/libgfortran.so.5
     29903:	  trying file=tls/haswell/x86_64/libgfortran.so.5
     29903:	  trying file=tls/haswell/libgfortran.so.5
     29903:	  trying file=tls/x86_64/libgfortran.so.5
     29903:	  trying file=tls/libgfortran.so.5
     29903:	  trying file=haswell/x86_64/libgfortran.so.5
     29903:	  trying file=haswell/libgfortran.so.5
     29903:	  trying file=x86_64/libgfortran.so.5
     29903:	  trying file=libgfortran.so.5
     29903:	  trying file=/home/user/ThirdParty-HSL/coinhsl/lib/libgfortran.so.5
     29903:	 search cache=/etc/ld.so.cache
     29903:	  trying file=/lib/x86_64-linux-gnu/libgfortran.so.5
     29903:	
     29903:	find library=libquadmath.so.0 [0]; searching
     29903:	 search path=glibc-hwcaps/x86-64-v3:glibc-hwcaps/x86-64-v2:tls/haswell/x86_64:tls/haswell:tls/x86_64:tls:haswell/x86_64:haswell:x86_64::/home/user/ThirdParty-HSL/coinhsl/lib		(LD_LIBRARY_PATH)
     29903:	  trying file=glibc-hwcaps/x86-64-v3/libquadmath.so.0
     29903:	  trying file=glibc-hwcaps/x86-64-v2/libquadmath.so.0
     29903:	  trying file=tls/haswell/x86_64/libquadmath.so.0
     29903:	  trying file=tls/haswell/libquadmath.so.0
     29903:	  trying file=tls/x86_64/libquadmath.so.0
     29903:	  trying file=tls/libquadmath.so.0
     29903:	  trying file=haswell/x86_64/libquadmath.so.0
     29903:	  trying file=haswell/libquadmath.so.0
     29903:	  trying file=x86_64/libquadmath.so.0
     29903:	  trying file=libquadmath.so.0
     29903:	  trying file=/home/user/ThirdParty-HSL/coinhsl/lib/libquadmath.so.0
     29903:	 search cache=/etc/ld.so.cache
     29903:	  trying file=/lib/x86_64-linux-gnu/libquadmath.so.0
     29903:	
     29903:	/home/user/ThirdParty-HSL/coinhsl/lib/libhsl.so: error: symbol lookup error: undefined symbol: isamax_ (fatal)
Exception of type: DYNAMIC_LIBRARY_FAILURE in file "/work/build/external_projects/src/ipopt-external/src/Common/IpLibraryLoader.cpp" at line 72:
 Exception message: /home/user/ThirdParty-HSL/coinhsl/lib/libhsl.so: undefined symbol: isamax_

EXIT: Library loading failure.
      solver  :   t_proc      (avg)   t_wall      (avg)    n_eval
       total  |  13.17ms ( 13.17ms) 855.89us (855.89us)         1


the file 'libhsl.so' exists after I made the symbolic link, but the library continues to fail loading
from the previous output I can only see one reason 

Exception message: /home/user/ThirdParty-HSL/coinhsl/lib/libhsl.so: undefined symbol: isamax_
some one suggested that its Lapack error. after using "ldd /home/user/ThirdParty-HSL/coinhsl/lib/libcoinhsl.so"
I got the output

linux-vdso.so.1 (0x00007ffd45975000)
libgfortran.so.5 => /lib/x86_64-linux-gnu/libgfortran.so.5 (0x00007f74f8400000)
libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f74f87df000)
libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f74f8000000)
libquadmath.so.0 => /lib/x86_64-linux-gnu/libquadmath.so.0 (0x00007f74f8797000)
libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f74f8777000)
/lib64/ld-linux-x86-64.so.2 (0x00007f74f8926000)

it seems that it does not list lapack even it was installed.
I have been sticking with this problem for more than two weeks, and now don't know what to do more.

**Reply by itbellix** [reply]
Hi @wissamkafa, did you manage to solve this? If so, could you share how?
Thanks!
